Classes/CharacterManager.gd

```diff
@@ -0,0 +1,79 @@
+class_name CharacterManager
+const script_name := "character_manager"
+
+static func create(character_name: String):
+	return CharacterDefaults.new().characters[character_name]
+
+
+static func load_class(character_data: Dictionary, character_class: String):
+	# Add equipment to the character inventory and equip those items
+	for item in CharacterDefaults.starting_equipment[character_class]:
+		var item_index = add(character_data, Items.items[item.name], item.quantity)
+		equip(character_data, item_index)
+	
+	# Add character attacks
+	for attack_type in CharacterDefaults.starting_attacks[character_class].keys():
+		for attack_name in CharacterDefaults.starting_attacks[character_class][attack_type]:
+			learn_attack(character_data, attack_type, attack_name)
+	
+	return character_data
+
+
+static func learn_attack(character_data: Dictionary, attack_type: String, attack_name: String):
+	character_data.attacks[attack_type].append(attack_name)
+
+
+static func calcuate_item_buffs(character_data):
+	Core.emit_signal("msg", "This function is not implemented yet!", Log.WARN, "character_manager")
+	#for item in player.items:
+	#	pass
+
+
+static func add(character_data: Dictionary, item_data: Dictionary, quantity:=1):
+	return InventoryManager.add(character_data.items.inventory, item_data, quantity)
+
+
+static func remove(character_data: Dictionary, item_data: Dictionary, quantity:=1):
+	InventoryManager.remove(character_data.items.inventory, item_data, quantity)
+
+
+static func equip(character_data: Dictionary, item_index: int):
+	if !character_data.items.inventory[item_index]:
+		return false
+	var item = character_data.items.inventory[item_index]
+
+	if item.level_requirement <= character_data.level:
+		character_data.items[item.type][item.subtype].append(item_index)
+	else:
+		return false
+
+
+static func unequip(character_data: Dictionary, item_index: int):
+	if !character_data.items.inventory[item_index]:
+		return false
+	var item = character_data.items.inventory[item_index]
+	
+	character_data.items[item.type][item.subtype].erase(item_index)
+
+static func set_level(character_data: Dictionary, level: int):
+	Core.emit_signal("msg", "Setting character level to " + str(character_data.level), Log.INFO, "character_manager")
+	character_data.level = level
+
+static func process_level(character_data: Dictionary):
+	Core.emit_signal("msg", "Processing character level " + str(character_data.level) + " stats", Log.INFO, "character_manager")
+	var temp_level = character_data.level
+	var count = 1
+	var new_character_data = character_data
+	while temp_level >= 5:
+		new_character_data.stats.strength += 1
+		new_character_data.stats.intelligence += 1
+		new_character_data.stats.agility +=1
+		new_character_data.stats.luck += 1
+		
+		if count%3==0:
+			new_character_data.stats.perception += 1
+			new_character_data.stats.endurance += 1
+			new_character_data.stats.charisma += 1
+	count+=1
+	temp_level-=5
+	return new_character_data.stats
```

Comment:
um, this won't work poroperly

step 1. level buffs had to be it's own stats like equip buffs
step 2. level buffs are cleared before calculating
step 3. the new value is created

Also side note, a class value needs to be given because that leveling is only for the hound.

Reply:
woops!

Reply:
Ok I think I fixed it!
```
{
  "meta": {
    "name": "Alrune",
    "nickname": "",
    "info": ""
  },
  "file": "Alrune - 1592107128",
  "stats": {
    "strength": 3,
    "perception": 3,
    "endurance": 3,
    "charisma": 3,
    "intelligence": 3,
    "agility": 3,
    "luck": 3,
    "melee": 0,
    "mana": 0,
    "defense": 0
  },
  "picture": {
    "path": "Friendlies/Tex_AnimeAva_28.png",
    "border": {
      "path": "ImageBorder.png",
      "shown": true
    }
  },
  "attacks": {
    "melee": [
      "punch"
    ],
    "ranged": [
      "quick_shot",
      "strike"
    ],
    "mana": []
  },
  "level": 10,
  "skills": {},
  "ap": {
    "max": 1,
    "current": 0.5
  },
  "health": {
    "max": 100,
    "current": 100
  },
  "mana": {
    "max": 50,
    "current": 100
  },
  "kills": 0,
  "items": {
    "armour": {
      "head": [],
      "torso": [],
      "arms": [],
      "legs": [],
      "feet": [],
      "shield": []
    },
    "equipment": {
      "melee": [
        2
      ],
      "ranged": [
        0
      ],
      "consumables": [
        1
      ],
      "magic": []
    },
    "inventory": [
      {
        "nickname": "Hunting bow",
        "name": "bow",
        "type": "equipment",
        "subtype": "ranged",
        "hand": "two-handed",
        "tags": [],
        "buffs": {
          "strength": 0,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 2,
          "luck": 0,
          "melee": 0,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 1
      },
      {
        "nickname": "Arrow",
        "name": "arrow",
        "type": "equipment",
        "subtype": "consumables",
        "hand": "one-handed",
        "tags": [],
        "buffs": {
          "strength": 0,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 0,
          "luck": 1,
          "melee": 4,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 15
      },
      {
        "nickname": "Dagger",
        "name": "dagger",
        "type": "equipment",
        "subtype": "melee",
        "hand": "one-handed",
        "tags": [],
        "buffs": {
          "strength": 0,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 0,
          "luck": 1,
          "melee": 5,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 1
      }
    ]
  },
  "nickname": "fpg",
  "level_stats": {
    "strength": 2,
    "perception": 0,
    "endurance": 0,
    "charisma": 0,
    "intelligence": 2,
    "agility": 2,
    "luck": 2,
    "melee": 0,
    "mana": 0,
    "defense": 0
  }
}
```

Reply:
Do we need to save this: "file": "Alrune - 1592107128"?
Wouldn't it just be saved by the environment not the player

Reply:
wasn't equipment and items just meant to be inventory.new()?

Reply:
Or is this more like a save data to load a character from?

Reply:
See commit d15061d38c7f4b452611f5fe3e5457b7dc30d1ae

(HEAD -> develop, origin/develop, github/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 17:36:31 2020 +1000

    refactor: hard code in class attributes
    
    - Add const starting_attacks to CharacterDefaults.gd
    - Item buffs are now calcuated when needed and not stored (no reduct
    complexity)
    - Items and Inventories are now stored as pure data
    - Functions moved to InventoryManager and ItemManager
    - Refactored InventoryManager
    - Updated item buffs to correct values
    - Implement function load_class(character, character_class)
    - Changed 'weapon' type for items to a more general 'equipment' type
    - Replaced CharacterCreation's match chosen_equip with
    CharacterManager.load_class(player, chosen_equip)

Reply:
atm, the 'file' attribute is used to save the character :D